File: solutions/python/trees/most-frequent-subtree-sum.py
# LeetCode 508 — Most Frequent Subtree Sum (Medium)
# Category: Trees · Approach: Postorder
# Time: O(n) | Space: O(n)
# Source: https://leetcode.com/problems/most-frequent-subtree-sum/

from collections import Counter

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val, self.left, self.right = val, left, right

def find_frequent_tree_sum(root: TreeNode) -> list[int]:
    cnt = Counter()
    def dfs(n):
        if not n:
            return 0
        s = n.val + dfs(n.left) + dfs(n.right)
        cnt[s] += 1
        return s
    dfs(root)
    if not cnt:
        return []
    hi = max(cnt.values())
    return [s for s, c in cnt.items() if c == hi]

File: solutions/typescript/trees/most-frequent-subtree-sum.ts
// LeetCode 508 — Most Frequent Subtree Sum (Medium)
// Category: Trees · Approach: Postorder
// Time: O(n) | Space: O(n)
// Source: https://leetcode.com/problems/most-frequent-subtree-sum/

class TreeNode {
  val: number;
  left: TreeNode | null = null;
  right: TreeNode | null = null;
  constructor(val = 0) { this.val = val; }
}

function findFrequentTreeSum(root: TreeNode | null): number[] {
  const cnt = new Map<number, number>();
  const dfs = (n: TreeNode | null): number => {
    if (!n) return 0;
    const s = n.val + dfs(n.left) + dfs(n.right);
    cnt.set(s, (cnt.get(s) ?? 0) + 1);
    return s;
  };
  dfs(root);
  let hi = 0;
  for (const c of cnt.values()) hi = Math.max(hi, c);
  const res: number[] = [];
  for (const [s, c] of cnt) if (c === hi) res.push(s);
  return res;
}
